api/src/main/java/jakarta/servlet/ServletContext.java

```diff
@@ -1436,4 +1517,87 @@
      */
     public void setResponseCharacterEncoding(String encoding);
+
+    /**
+     * The mode in which this context handles the encoding and decoding of URI paths. This mode affects the behavior of:
+     * <ul>
+     * <li>{@link HttpServletRequest#getContextPath()}</li>
+     * <li>{@link HttpServletRequest#getServletPath()}</li>
+     * <li>{@link HttpServletRequest#getPathInfo()}</li>
+     * <li>{@link HttpServletRequest#getPath()}</li>
+     * <li>{@link HttpServletRequest#getRealPath(String)}</li>
+     * <li>{@link HttpServletRequest#getRequestDispatcher(String)}</li>
+     * <li>{@link ServletContext#getContextPath()}</li>
+     * <li>{@link ServletContext#getContext(String)}</li>
+     * <li>{@link ServletContext#getRequestDispatcher(String)}</li>
+     * <li>{@link ServletContext#getRealPath(String)}</li>
+     * <li>{@link ServletContext#getPath(String)}</li>
+     * <li>{@link ServletContext#getResource(String)}</li>
+     * <li>{@link ServletContext#getResourceAsStream(String)}</li>
+     * <li>{@link ServletContext#getResourceAsPath(String)}</li>
+     * <li>{@link ServletContext#getResourcePaths(String)}</li>
+     * <li>{@link AsyncContext#dispatch(String)}</li>
+     * <li>{@link AsyncContext#dispatch(ServletContext, String)}</li>
+     * </ul>
+     * for the purposes of these modes:
+     * <ul>
+     * <li>a 'reserved character' is a character defined as <code>reserved</code> by
+     * <a href="https://datatracker.ietf.org/doc/html/rfc3986#section-2.2">RFC3986 Section 2.2</a> or the ';'
+     * character.</li>
+     * <li>a 'path parameter' is the part of a URI segment as defined by the obsoleted
+     * <a href="https://datatracker.ietf.org/doc/html/rfc2396#section-3.3">RFC2396 Section 3.3</a>.</li>
+     * </ul>
+     */
+    public enum PathCodingMode {
+        /**
+         * Effected methods will return/accept paths as was done prior to the introduction of the PathCodingMode in Servlet 6.0.
+         * The path returned by {@link HttpServletRequest#getContextPath()} is not decoded. Paths returned by the methods
+         * {@link HttpServletRequest#getServletPath()} and {@link HttpServletRequest#getPathInfo()} are fully decoded and have
+         * all path parameters removed. Decoding may include ambiguous characters such as a decoded <code>%2F</code>, which is
+         * indistinguishable from a segment separator '/' in the request URI. Methods accepting paths may accept encoded
+         * characters. A '?' character in a path will be treated as indicating a query string and will not be encoded. Some
+         * implementation specific behavioural differences may be apparent in this mode.
+         */
+        LEGACY,
+
+        /**
+         * Effected methods will only return/accept paths that do not contain any encoded reserved characters and with all path
+         * parameters removed. If the requested URI contains encoded reserved characters, then a call to a method that returns
```

Comment:
Shouldn't this just return bad request and be done with it? Historically these methods (e.g. `getPathInfo`) have never thrown an exception, so changing this would potentially have unexpected consequences (especially as this can be triggered by a remote user). If an app is not using try/finally properly this could potentially result in a resource leak, and the possibility to DOS a poorly coded app by sending requests with reserved characters.

Reply:
I'd be ok with this (it is essentially what Jetty already does unless you configure exceptions for strange looking URIs).

My only concern is that apps that currently do use `getRequestURI` and do their own handling will need to be modified/configured to set the `ENCODED` mode rather than just working out of the box.      But there is no way to know a-priori if an application will call the affected APIs or not, so either we delay enforcement and throw (which may have some problems) or block early with a 400 (may require some config for existing apps)

Reply:
Won't existing apps be using LEGACY anyway?

Reply:
I like the alternative.

Reply:
done